qualtran/bloqs/chemistry/quad_fermion/givens_bloq.py

```diff
@@ -0,0 +1,218 @@
+    the phase gradient state which which is $2(b_{\mathrm{grad}}-2)$ where $b_{\mathrm{grad}}$
+    is the size of the phasegradient register.
+
+    References:
+        [Compilation of Fault-Tolerant Quantum Heuristics for Combinatorial Optimization](https://arxiv.org/abs/2007.07391).
+        Section II-C: Oracles for phasing by cost function. Appendix A: Addition for controlled rotations
+    Args:
+        phasegrad_bitsize int: size of phase gradient which is also the size of the register
+                               representing the binary fraction of the rotation angle
+
+
+    """
+    phasegrad_bitsize: int
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(
+            target_i=QBit(),
+            target_j=QBit(),
+            rom_data=QFxp(self.phasegrad_bitsize, self.phasegrad_bitsize, signed=False),
+            phase_gradient=QFxp(self.phasegrad_bitsize, self.phasegrad_bitsize, signed=False),
+        )
+
+    def build_composite_bloq(
+        self,
+        bb: BloqBuilder,
+        target_i: SoquetT,
+        target_j: SoquetT,
+        rom_data: SoquetT,
+        phase_gradient: SoquetT,
+    ) -> Dict[str, SoquetT]:
+        # set up rz-rotation via phase-gradient state
+        add_into_phasegrad_gate = AddIntoPhaseGrad(
+            x_bitsize=self.phasegrad_bitsize,
+            phase_bitsize=self.phasegrad_bitsize,
+            right_shift=0,
+            sign=1,
+            controlled=1,
+        )
+
+        # clifford block
+        target_i = bb.add(XGate(), q=target_i)
+        target_j = bb.add(SGate(), q=target_j)
+        target_j, target_i = bb.add(CNOT(), ctrl=target_j, target=target_i)
+        target_j = bb.add(Hadamard(), q=target_j)
+        target_i = bb.add(SGate(is_adjoint=True), q=target_i)
+        target_j, target_i = bb.add(CNOT(), ctrl=target_j, target=target_i)
+
+        # parallel rz (Can probably be improved with single out of place adder into a single ancilla
+        target_i, rom_data, phase_gradient = bb.add(
+            add_into_phasegrad_gate, x=rom_data, phase_grad=phase_gradient, ctrl=target_i
+        )
+        target_j, rom_data, phase_gradient = bb.add(
+            add_into_phasegrad_gate, x=rom_data, phase_grad=phase_gradient, ctrl=target_j
+        )
+
+        # clifford block
+        target_j, target_i = bb.add(CNOT(), ctrl=target_j, target=target_i)
+        target_i = bb.add(XGate(), q=target_i)
+        target_i, target_j = bb.add(CNOT(), ctrl=target_i, target=target_j)
+        target_j = bb.add(Hadamard(), q=target_j)
+        target_j, target_i = bb.add(CNOT(), ctrl=target_j, target=target_i)
+        target_i = bb.add(SGate(), q=target_i)
+
+        return {
+            'target_i': target_i,
+            'target_j': target_j,
+            'rom_data': rom_data,
+            'phase_gradient': phase_gradient,
+        }
+
+    # def _t_complexit_(self)-> 'TComplexity':
```

Comment:
Remove the commented out code. I can't remember if t_complexity is required if you have the decomposition.